// pages/create-article/index.js
import React, { useState, useEffect, useContext } from 'react';
import { Redirect } from 'react-router-dom';

import { ArticleForm } from 'components/article-form';
import useFetch from 'hooks/useFetch';
import { CurrentUserContext } from 'contexts/currentUser';

export const CreateArticle = () => {
  const apiUrl = '/articles';
  const [{response, error}, doFetch] = useFetch(apiUrl);
  const [isSuccessfullSubmit, setIsSuccessfullSubmit] = useState(false);
  const [currentUserState] = useContext(CurrentUserContext);

  const initialValues = {
    title: '',
    description: '',
    body: '',
    tagList: [],
  };

  const handleSubmit = article => {
    doFetch({
      method: 'post',
      data: {
        article,
      }
    });
  };

  useEffect(() => {
    if (!response) {
      return;
    }

    setIsSuccessfullSubmit(true);
  }, [setIsSuccessfullSubmit, response]);

  if (currentUserState.isLoggedIn === false) {
    return (
      <Redirect to='/' />
    );
  }

  if (isSuccessfullSubmit) {
    return (
      <Redirect to={`/articles/${response.article.slug}`} />
    );
  }

  return (
    <div>
      <ArticleForm
        errors={((error && error.errors) || {})}
        initialValues={initialValues}
        onSubmit={handleSubmit}
      />
    </div>
  );
}
